src/audio-sdlmixer.cc

```diff
@@ -0,0 +1,291 @@
+  float volume = get_volume();
+  set_volume(volume + 0.1f);
+}
+
+void
+audio_sdlmixer_t::volume_down() {
+  float volume = get_volume();
+  set_volume(volume - 0.1f);
+}
+
+audio_track_t *
+sfx_player_t::create_track(int track_id) {
+  size_t size = 0;
+  void *data = data_get_object(DATA_SFX_BASE + track_id, &size);
+  if (data == NULL) {
+    return NULL;
+  }
+
+  void *wav = sfx2wav(data, size, &size, -0x20);
+  if (wav == NULL) {
+    return NULL;
+  }
+
+  SDL_RWops *rw = SDL_RWFromMem(wav, static_cast<int>(size));
+  Mix_Chunk *chunk = Mix_LoadWAV_RW(rw, 0);
+  free(wav);
+  if (chunk == NULL) {
+    LOGE("audio-sdlmixer", "Mix_LoadWAV_RW: %s.", Mix_GetError());
+    return NULL;
+  }
+
+  return new sfx_track_t(chunk);
+}
+
+void
+sfx_player_t::enable(bool enable) {
+  enabled = enable;
+  if (!enabled) {
+    stop();
+  }
+}
+
+void
+sfx_player_t::stop() {
+  Mix_HaltChannel(-1);
+}
+
+float
+sfx_player_t::get_volume() {
+  return static_cast<float>(Mix_Volume(-1, -1)) /
+         static_cast<float>(MIX_MAX_VOLUME);
+}
+
+void
+sfx_player_t::set_volume(float volume) {
+  volume = std::max(std::min(0.f, volume), 1.f);
+  float mix_volume = static_cast<float>(MIX_MAX_VOLUME) * volume;
+  Mix_Volume(-1, static_cast<int>(mix_volume));
+}
+
+void
+sfx_player_t::volume_up() {
+  set_volume(get_volume() + 0.1f);
+}
+
+void
+sfx_player_t::volume_down() {
+  set_volume(get_volume() - 0.1f);
+}
+
+sfx_track_t::sfx_track_t(Mix_Chunk *chunk) {
+  this->chunk = chunk;
+}
+
+sfx_track_t::~sfx_track_t() {
+  Mix_FreeChunk(chunk);
+}
+
+void
+sfx_track_t::play() {
+  int r = Mix_PlayChannel(-1, chunk, 0);
+  if (r < 0) {
+    LOGE("audio-sdlmixer", "Could not play SFX clip: %s.", Mix_GetError());
+  }
+}
+
+audio_track_t *
+midi_player_t::create_track(int track_id) {
+  size_t size = 0;
+  void *data = data_get_object(DATA_MUSIC_GAME + track_id, &size);
+  if (NULL == data) {
+    return NULL;
+  }
+
+  void *midi = xmi2mid(data, size, &size);
+  SDL_RWops *rw = SDL_RWFromMem(midi, static_cast<int>(size));
+  Mix_Music *music = Mix_LoadMUS_RW(rw, 0);
+  free(midi);
+  if (music == NULL) {
+    return NULL;
+  }
+
+  return new midi_track_t(music);
+}
+
+void
+midi_player_t::enable(bool enable) {
+  enabled = enable;
+  if (!enabled) {
+    stop();
+  }
+}
+
+void
+midi_player_t::stop() {
+  Mix_HaltMusic();
+}
+
+float
+midi_player_t::get_volume() {
+  return static_cast<float>(Mix_VolumeMusic(-1)) /
+         static_cast<float>(MIX_MAX_VOLUME);
+}
+
+void
+midi_player_t::set_volume(float volume) {
+  volume = std::max(std::min(0.f, volume), 1.f);
+  float mix_volume = static_cast<float>(MIX_MAX_VOLUME) * volume;
+  Mix_VolumeMusic(static_cast<int>(mix_volume));
+}
+
+void
+midi_player_t::volume_up() {
+  set_volume(get_volume() + 0.1f);
+}
+
+void
+midi_player_t::volume_down() {
+  set_volume(get_volume() - 0.1f);
+}
+
+/*
+static void
+midi_track_finished() {
+  if (midi_enabled) {
+    midi_play_track(current_track);
+  }
+}
+*/
```

Comment:
It's my fault. I decided that porting as is impractical, but forgot to do well (: